In [12]:
# Loading the dataset using pandas"

import pandas as pd

df = pd.read_csv('customer_shopping_behavior.csv')

In [13]:
df.head()

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually


### Analysis
- `df.head()` displays the first 5 rows of the DataFrame.
- The dataset contains customer information, purchase details, product details, and shopping behavior.
- Key columns include **Customer ID, Age, Gender, Item Purchased, Category, Purchase Amount, Location, Season, Review Rating, Subscription Status, Discount Applied, Payment Method,** and **Frequency of Purchases**.


In [14]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   str    
 3   Item Purchased          3900 non-null   str    
 4   Category                3900 non-null   str    
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   str    
 7   Size                    3900 non-null   str    
 8   Color                   3900 non-null   str    
 9   Season                  3900 non-null   str    
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   str    
 12  Shipping Type           3900 non-null   str    
 13  Discount Applied        3900 non-null   str    
 14  Promo Code Used         3900 non-null   str    
 15

### Analysis

- The dataset contains **3,900 rows and 18 columns**.
- Most columns have **3,900 non-null values**, indicating there are no missing values in those columns.
- The **Review Rating** column has **3,863 non-null values**, meaning **37 values are missing**.
- The dataset contains **numeric columns** such as Age, Purchase Amount, Previous Purchases, and Review Rating.
- The remaining columns are **categorical/string columns** containing information such as Gender, Category, Location, Season, Payment Method, and Subscription Status.
- Since only `Review Rating` contains missing values, it will need to be handled before performing analysis involving customer ratings.

In [15]:
# Summary statistics using .describe()
df.describe(include='all')

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
count,3900.000000,3900.000000,3900,3900,3900,3900.000000,3900,3900,3900,3900,3863.000000,3900,3900,3900,3900,3900.000000,3900,3900
unique,NaN,NaN,2,25,4,NaN,50,4,25,4,NaN,2,6,2,2,NaN,6,7
top,NaN,NaN,Male,Blouse,Clothing,NaN,Montana,M,Olive,Spring,NaN,No,Free Shipping,No,No,NaN,PayPal,Every 3 Months
freq,NaN,NaN,2652,171,1737,NaN,96,1755,177,999,NaN,2847,675,2223,2223,NaN,677,584
mean,1950.500000,44.068462,NaN,NaN,NaN,59.764359,NaN,NaN,NaN,NaN,3.750065,NaN,NaN,NaN,NaN,25.351538,NaN,NaN
std,1125.977353,15.207589,NaN,NaN,NaN,23.685392,NaN,NaN,NaN,NaN,0.716983,NaN,NaN,NaN,NaN,14.447125,NaN,NaN
min,1.000000,18.000000,NaN,NaN,NaN,20.000000,NaN,NaN,NaN,NaN,2.500000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN
25%,975.750000,31.000000,NaN,NaN,NaN,39.000000,NaN,NaN,NaN,NaN,3.100000,NaN,NaN,NaN,NaN,13.000000,NaN,NaN
50%,1950.500000,44.000000,NaN,NaN,NaN,60.000000,NaN,NaN,NaN,NaN,3.800000,NaN,NaN,NaN,NaN,25.000000,NaN,NaN
75%,2925.250000,57.000000,NaN,NaN,NaN,81.000000,NaN,NaN,NaN,NaN,4.400000,NaN,NaN,NaN,NaN,38.000000,NaN,NaN


### Analysis

- The dataset contains **3,900 customer records** across **18 columns**.
- The **Age** ranges from **18 to 70 years**, with an average age of approximately **44 years**.
- The **Purchase Amount** ranges from **$20 to $100**, with an average purchase value of approximately **$59.76**.
- **Review Rating** ranges from **2.5 to 5.0**, with an average rating of approximately **3.75**. There are **37 missing ratings**.
- **Previous Purchases** range from **1 to 50**, with an average of approximately **25 purchases** per customer.
- Among categorical variables, **Male** is the most frequent gender, **Blouse** is the most frequently purchased item, and **Clothing** is the most common category.
- **Montana** is the most frequent location, while **M** is the most common size.
- **Spring** is the most frequent season in the dataset.
- Most customers have **no subscription**, with **2,847 customers** marked as `No`.
- **Free Shipping** is the most common shipping type, occurring **675 times**.
- **Discount Applied** and **Promo Code Used** both have `No` as the most frequent value, with **2,223 records** each.
- **PayPal** is the most frequently used payment method, with **677 transactions**.
- **Every 3 Months** is the most common purchase frequency, with **584 customers**.
- Overall, the dataset contains a mix of **customer demographics, purchasing behavior, product information, and transaction details**, making it suitable for detailed customer and sales analysis.

In [16]:
# Checking if missing data or null values are present in the dataset

df.isnull().sum()

Customer ID                0
Age                        0
Gender                     0
Item Purchased             0
Category                   0
Purchase Amount (USD)      0
Location                   0
Size                       0
Color                      0
Season                     0
Review Rating             37
Subscription Status        0
Shipping Type              0
Discount Applied           0
Promo Code Used            0
Previous Purchases         0
Payment Method             0
Frequency of Purchases     0
dtype: int64

### Analysis

- `df.isnull().sum()` is used to check the number of missing values in each column.
- The dataset contains **37 missing values in the `Review Rating` column**.
- All other columns have **0 missing values**, indicating that they are complete.
- Since the missing values are limited to one column, we can handle them separately without affecting the rest of the dataset.
- Before performing analysis involving `Review Rating`, the missing values should be appropriately handled.

In [17]:
# Imputing missing values in Review Rating column with the median rating of the product category

df['Review Rating'] = df.groupby('Category')['Review Rating'].transform(lambda x: x.fillna(x.median()))

### Analysis

- The missing values in the `Review Rating` column are being handled using **median imputation**.
- The median rating is calculated **separately for each product category** using `groupby('Category')`.
- Missing ratings are then replaced with the **median rating of their respective category**.
- This approach is better than using the overall median because customers in different categories may have different typical rating patterns.
- `transform()` ensures that the category-level median is aligned with each original row while keeping the **same number of rows** in the DataFrame.
- Using the median also helps reduce the influence of unusually high or low ratings.

In [18]:
df.isnull().sum()

Customer ID               0
Age                       0
Gender                    0
Item Purchased            0
Category                  0
Purchase Amount (USD)     0
Location                  0
Size                      0
Color                     0
Season                    0
Review Rating             0
Subscription Status       0
Shipping Type             0
Discount Applied          0
Promo Code Used           0
Previous Purchases        0
Payment Method            0
Frequency of Purchases    0
dtype: int64

In [19]:
# Renaming columns according to snake casing for better readability and documentation

df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ','_')
df = df.rename(columns={'purchase_amount_(usd)':'purchase_amount'})

### Analysis

- The column names are standardized to **lowercase** to make them easier to read and work with.
- Spaces in column names are replaced with **underscores (`_`)**, following the **snake_case naming convention**.
- The column `purchase_amount_(usd)` is renamed to **`purchase_amount`** for simplicity.
- Standardized column names make the dataset easier to use in **Python, SQL, and data visualization**.
- This also improves code readability and avoids having to use spaces or special characters when referencing columns.

In [20]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='str')

In [21]:
bins = [0, 25, 35, 50, 100]
labels = ['Young Adult', 'Adult', 'Middle-aged', 'Senior']

df['age_group'] = pd.cut(
    df['age'],
    bins=bins,
    labels=labels
)

### Analysis

- The `age` column is divided into **four age groups** using `pd.cut()`.
- The defined age ranges are:
  - **0–25:** Young Adult
  - **25–35:** Adult
  - **35–50:** Middle-aged
  - **50–100:** Senior
- A new column called **`age_group`** is created based on these age ranges.
- Converting individual ages into groups makes it easier to analyze and compare **purchasing behavior across different age segments**.
- These age groups can later be used for **groupby analysis, visualizations, and customer segmentation**.

In [22]:
df[['age','age_group']].head(10)

,age,age_group
0,55,Senior
1,19,Young Adult
2,50,Middle-aged
3,21,Young Adult
4,45,Middle-aged
5,46,Middle-aged
6,63,Senior
7,27,Adult
8,26,Adult
9,57,Senior


### Analysis

- The output shows the original `age` values along with their corresponding `age_group`.
- This confirms that each customer's age has been successfully categorized into the appropriate age segment.
- The `age_group` column can now be used to compare **customer behavior, purchase amounts, and preferences across different age groups**.
- Grouping customers by age makes the dataset easier to analyze and can help identify **age-based purchasing patterns**.

In [ ]:
# create new column purchase_frequency_days

frequency_mapping = {
    'Fortnightly': 14,
    'Weekly': 7,
    'Monthly': 30,
    'Quarterly': 90,
    'Bi-Weekly': 14,
    'Annually': 365,
    'Every 3 Months': 90
}

df['purchase_frequency_days'] = df['frequency_of_purchases'].map(frequency_mapping)

### Analysis

- A new column, **`purchase_frequency_days`**, is created to convert purchase frequency categories into numerical values.
- Each frequency is mapped to the approximate number of days between purchases.
- For example, **Weekly = 7 days**, **Monthly = 30 days**, **Quarterly = 90 days**, and **Annually = 365 days**.
- `map()` is used to replace each categorical frequency with its corresponding number of days.
- Converting purchase frequency into a numerical format makes it easier to perform **calculations, comparisons, customer segmentation, and frequency-based analysis**.

In [13]:
df[['purchase_frequency_days','frequency_of_purchases']].head(10)

,purchase_frequency_days,frequency_of_purchases
0,14,Fortnightly
1,14,Fortnightly
2,7,Weekly
3,7,Weekly
4,365,Annually
5,7,Weekly
6,90,Quarterly
7,7,Weekly
8,365,Annually
9,90,Quarterly


### Analysis

- The output confirms that the `frequency_of_purchases` categories have been successfully converted into numerical values in `purchase_frequency_days`.
- For example, **Fortnightly = 14 days**, **Weekly = 7 days**, **Quarterly = 90 days**, and **Annually = 365 days**.
- This confirms that the mapping has been applied correctly to the displayed records.
- The new numerical column can now be used for **calculations, comparisons, and customer purchasing frequency analysis**.

In [14]:
df[['discount_applied','promo_code_used']].head(10)

,discount_applied,promo_code_used
0,Yes,Yes
1,Yes,Yes
2,Yes,Yes
3,Yes,Yes
4,Yes,Yes
5,Yes,Yes
6,Yes,Yes
7,Yes,Yes
8,Yes,Yes
9,Yes,Yes


In [15]:
(df['discount_applied'] == df['promo_code_used']).all()

np.True_

### Analysis

- The output shows that `discount_applied` and `promo_code_used` have the same values for the displayed records.
- The `.all()` check confirms that **both columns have exactly the same values across all 3,900 records**.
- Since the result is `True`, there is **no difference between the two columns** in this dataset.
- Therefore, both columns contain duplicate information, and one of them could be **removed during data cleaning** to avoid redundancy.

In [16]:
# Dropping promo code used column

df = df.drop('promo_code_used', axis=1)

### Analysis

- The `promo_code_used` column is removed from the DataFrame using `drop()`.
- This column was removed because it contained the **same information as `discount_applied`**.
- Removing the duplicate column helps **reduce redundancy** and keeps the dataset cleaner.
- The `discount_applied` column is retained as it provides the required information about whether a discount was applied.

In [24]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases', 'age_group'],
      dtype='str')

## Connecting Python script to PostgreSQL

In [25]:
!pip install psycopg2-binary sqlalchemy

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.1.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [20]:
from sqlalchemy import create_engine

# Step 1: Connect to PostgreSQL
# Replace placeholders with your actual details
username = "postgres"      # default user
password = "root123" # the password you set during installation
host = "localhost"         # if running locally
port = "5433"              # default PostgreSQL port
database = "customer_behavior"    # the database you created in pgAdmin

engine = create_engine(f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}")

# Step 2: Load DataFrame into PostgreSQL
table_name = "customer"   # choose any table name
df.to_sql(table_name, engine, if_exists="replace", index=False)

print(f"Data successfully loaded into table '{table_name}' in database '{database}'.")

Data successfully loaded into table 'customer' in database 'customer_behavior'.


### Analysis

- `create_engine()` is used to establish a connection between Python and the **PostgreSQL database** using SQLAlchemy.
- The database connection is configured with the **username, password, host, port, and database name**.
- `df.to_sql()` is used to transfer the cleaned DataFrame from Python into PostgreSQL.
- The DataFrame is stored as a table named **`customer`** in the `customer_behavior` database.
- `if_exists='replace'` means that if a table named `customer` already exists, it will be **replaced with the current DataFrame**.
- `index=False` prevents the Pandas DataFrame index from being added as an unnecessary column in the PostgreSQL table.
- The printed message confirms that the data has been successfully loaded into PostgreSQL.
- This allows the cleaned dataset to be further analyzed using **SQL queries**.

## Code for MySQL

In [ ]:
!pip install pymysql sqlalchemy

In [ ]:
from sqlalchemy import create_engine

# MySQL connection
username = "root"
password = "your_password"
host = "localhost"
port = "3306"
database = "customer_behavior"

engine = create_engine(f"mysql+pymysql://{username}:{password}@{host}:{port}/{database}")

# Write DataFrame to MySQL
table_name = "customer"   # choose any table name
df.to_sql(table_name, engine, if_exists="replace", index=False)

# Read back sample
pd.read_sql("SELECT * FROM customer LIMIT 5;", engine)

## Code for MS SQL Server

In [ ]:
!pip install pyodbc sqlalchemy 

In [ ]:
# Install required libraries

from sqlalchemy import create_engine
from urllib.parse import quote_plus

# SQL Server connection
username = "sa"
password = "your_password"
host = "localhost"
port = "1433"
database = "customer_behavior"

# Note: requires Microsoft ODBC Driver installed separately on your machine
driver = quote_plus("ODBC Driver 17 for SQL Server")
engine = create_engine(f"mssql+pyodbc://{username}:{password}@{host},{port}/{database}?driver={driver}")

# Write DataFrame to SQL Server
df.to_sql("customer", engine, if_exists="replace", index=False)

# Read back sample (SQL Server uses TOP instead of LIMIT)
pd.read_sql("SELECT TOP 5 * FROM customer;", engine)